lib/src/json_serializable_generator.dart

```diff
@@ -228,196 +238,35 @@ class JsonSerializableGenerator
   /// representing the serialization of a value.
-  String _fieldToJsonMapValue(String expression, DartType fieldType,
-      [int depth = 0]) {
-    for (var helper in _typeHelpers) {
-      if (helper.canSerialize(fieldType)) {
-        return helper.serialize(fieldType, expression);
-      }
-    }
-
-    if (fieldType.isDynamic ||
-        fieldType.isObject ||
-        _stringBoolNumChecker.isAssignableFromType(fieldType)) {
-      return expression;
-    }
-
-    if (_coreIterableChecker.isAssignableFromType(fieldType)) {
-      return _listFieldToJsonMapValue(expression, fieldType, depth);
-    }
-
-    if (_coreMapChecker.isAssignableFromType(fieldType)) {
-      return _mapFieldToJsonMapValue(expression, fieldType, depth);
-    }
-
-    throw new _UnsupportedTypeError(fieldType, expression);
-  }
-
-  String _listFieldToJsonMapValue(
-      String expression, DartType fieldType, int depth) {
-    assert(_coreIterableChecker.isAssignableFromType(fieldType));
-
-    // This block will yield a regular list, which works fine for JSON
-    // Although it's possible that child elements may be marked unsafe
-
-    var isList = _coreListChecker.isAssignableFromType(fieldType);
-    var substitute = "v$depth";
-    var subFieldValue = _fieldToJsonMapValue(
-        substitute, _getIterableGenericType(fieldType), depth + 1);
-
-    // In the case of trivial JSON types (int, String, etc), `subFieldValue`
-    // will be identical to `substitute` – so no explicit mapping is needed.
-    // If they are not equal, then we to write out the substitution.
-    if (subFieldValue != substitute) {
-      // TODO: the type could be imported from a library with a prefix!
-      expression = "${expression}?.map(($substitute) => $subFieldValue)";
-
-      // expression now represents an Iterable (even if it started as a List
-      // ...resetting `isList` to `false`.
-      isList = false;
-    }
-
-    if (!isList) {
-      // If the static type is not a List, generate one.
-      expression += "?.toList()";
-    }
-
-    return expression;
-  }
-
-  String _mapFieldToJsonMapValue(
-      String expression, DartType fieldType, int depth) {
-    assert(_coreMapChecker.isAssignableFromType(fieldType));
-    var args = _getTypeArguments(fieldType, _coreMapChecker);
-    assert(args.length == 2);
-
-    var keyArg = args.first;
-    var valueType = args.last;
-
-    // We're not going to handle converting key types at the moment
-    // So the only safe types for key are dynamic/Object/String
-    var safeKey = keyArg.isDynamic ||
-        keyArg.isObject ||
-        _coreStringChecker.isExactlyType(keyArg);
-
-    var safeValue = valueType.isDynamic ||
-        valueType.isObject ||
-        _stringBoolNumChecker.isAssignableFromType(valueType);
+  String _serialize(DartType targetType, String expression) => _allHelpers
+      .map((th) => th.serialize(targetType, expression, _serialize))
```

Comment:
Maybe just `h` for 'helper' coming from `_allHelpers` -> or change it to `_allTypeHelpers` and keep `th`